### ISSUE TITLE
Distill SongFormer boundary/readiness and section-function knowledge into training-only auxiliary heads

### ISSUE THREAD

**Issue by rhythmatician** [rhythmatician/FractalSync#102]
Part of #77

Coordinates with #85 (offline transition ground truth), #104 (shared causal Player trunk / auxiliary-head contract), #95 (Julia ColorIntent), #103 (other offline machine-listening teachers), #109 (runtime AudioActivation), and #81/#87 (behavioral transition objectives/evaluation).

## Motivation

We want the learned Player to behave more like an experienced musician approaching a possible structural change:

> prepare when causal evidence makes a transition plausible, make reversible choices while uncertain, then commit when later causal evidence confirms the new regime.

That behavior should be learned in the **shared causal Player representation**, not hard-coded into `runtime-core` as a handcrafted `transition_readiness` feature and not supplied by SongFormer at runtime.

The original version of this ticket assumed the postprocessed `outputs_songformer/*.json` section boundaries were the only useful teacher signal. Reviewing upstream SongFormer shows a richer opportunity.

SongFormer has two learned output heads:

```text
boundary_head -> boundary_logits[t]
function_head -> function_logits[t, class]
```

The public inference path aggregates those dense logits over the song and only afterward postprocesses them into hard structural segments. Boundary logits are converted to sigmoid probabilities; function logits are converted to softmax probabilities; then boundary peak-picking and segment labeling produce the final `label/start/end` JSON.

Upstream references:

- model heads: https://github.com/ASLP-lab/SongFormer/blob/main/src/SongFormer/models/SongFormer.py
- inference / dense-logit aggregation: https://github.com/ASLP-lab/SongFormer/blob/main/src/SongFormer/infer/infer.py
- postprocessing: https://github.com/ASLP-lab/SongFormer/blob/main/src/SongFormer/postprocessing/functional.py
- label vocabulary: https://github.com/ASLP-lab/SongFormer/blob/main/src/SongFormer/dataset/label2id.py

The default public inference configuration is `SongForm-HX-8Class`, approximately:

```text
intro
verse
chorus
bridge
instrumental
outro
silence
pre-chorus
```

SongFormer is explicitly an **offline/noncausal teacher** for FractalSync. Its own inference/postprocessing may use whole-song context and future evidence. The Student must remain strictly causal.

## Core question

Can distilling SongFormer's **dense soft structural knowledge** into training-only auxiliary heads improve the causal Player trunk and, consequently, the actual Mandelbrot motion and Julia presentation chosen by PlayerPolicy?

Conceptually:

```text
                OFFLINE ONLY
full song -> SongFormer
              |
              +-> boundary soft timeline
              +-> section/function soft timeline
              +-> hard segment timeline
                         |
                         v
                teacher targets
                         |
                         v
runtime-causal PlayerObservation history
                         |
                         v
                  shared trunk h_t
                 /       |        \
                /        |         \
         Controls v2  readiness    section/function
                     aux head       aux head
                     TRAIN ONLY      TRAIN ONLY
```

Backpropagation from auxiliary losses should make structural information available to the same shared state that chooses both Mandelbrot Controls and Julia ViewControls.

The auxiliary outputs are not product outputs and may be removed entirely from the exported runtime model.

## Stage 0 — preserve the dense SongFormer teacher signal

The current FractalSync SongFormer workflow primarily retains postprocessed hard segments. Before testing auxiliary supervision, preserve a versioned dense teacher artifact per song.

At minimum retain, at SongFormer's native teacher frame times:

```text
boundary_logit[t]
boundary_probability[t]

function_logits[t, class]
function_probability[t, class]

postprocessed_segments[] = {label, start, end}

teacher metadata:
  SongFormer revision/checkpoint
  SongFormer config / label vocabulary
  MuQ/MusicFM dependency/checkpoint identity where practical
  native teacher frame rate/time semantics
  source-audio identity/hash
  generation timestamp/tool version
```

Requirements:

- preserve the **native dense output**, not only the final peak-picked segmentation;
- do not treat hard postprocessed labels as more certain than the underlying soft teacher evidence;
- retain enough metadata to regenerate/compare teacher timelines after a model/config change;
- align teacher time to FractalSync's canonical sample-clock timeline deterministically for training;
- keep the native teacher timestamps as source truth rather than pretending SongFormer runs at the canonical 48 kHz / 1024-hop rate;
- document interpolation/resampling semantics when teacher targets are consumed at Player ticks;
- no SongFormer dependency enters the browser/live runtime.

This ticket should implement the minimum SongFormer-specific artifact needed for the experiment. Do **not** build a large generic multi-teacher framework unless another concrete consumer demonstrates that abstraction is already needed.

## Stage 1 — boundary and transition-readiness distillation

This is the primary bet.

Let SongFormer provide a soft boundary signal

```text
z_B^T(t) = teacher boundary logit
p_B^T(t) = sigmoid(z_B^T(t)).
```

### A. Boundary-now auxiliary target

Train a small Student head from causal shared state `h_t`:

```text
p_B^S(t) = sigmoid(g_B(h_t))
```

with a soft-label BCE/cross-entropy objective against `p_B^T(t)`.

This asks whether the causal trunk can recognize that a structural boundary is occurring *now*.

### B. Transition-readiness / future-horizon targets

More importantly, derive training-only targets representing how much teacher boundary evidence lies in the **future** of Student time `t`.

Evaluate a deliberately small horizon set, initially something like:

```text
H in {0.5 s, 1 s, 2 s, 4 s}
```

and define targets conceptually as

```text
y_H^T(t) = teacher evidence that a structural boundary occurs within (t, t+H].
```

Candidate constructions include:

- maximum boundary probability inside the horizon;
- integrated / kernel-weighted boundary mass;
- a calibrated discrete-time hazard / survival target;
- a smooth time-to-next-boundary kernel derived from the dense teacher signal.

Do **not** assume framewise SongFormer boundary probabilities are statistically independent merely to turn them into a mathematically convenient probability. If a hazard/survival interpretation is used, calibrate/validate it explicitly.

The Student predicts only from causal state:

```text
y_H^S(t) = g_H(h_t),
```

so the training experiment asks:

> **How much of SongFormer's future-aware structural judgment is actually predictable from FractalSync's causal observation history?**

The target should encourage anticipation, not merely punish a model after the transition was already missed.

## Stage 2 — soft section/function distillation

Separately test whether SongFormer's section-role distribution improves the shared representation.

Let

```text
p_F^T(t) = softmax(z_F^T(t) / T)
p_F^S(t) = softmax(g_F(h_t) / T)
```

and use a temperature-scaled distillation loss such as

```text
L_F(t)
  = w_T(t) * T^2 * KL(p_F^T(t) || p_F^S(t)).
```

where `w_T(t)` may decrease when teacher entropy is high.

Important interpretation:

> **The purpose is not to make `VERSE / CHORUS / BRIDGE / ...` part of FractalSync's runtime ontology. These are temporary teacher coordinates used to shape the shared causal representation.**

This matters because conventional SongFormer section classes may fit some music poorly or ambiguously. Preserve uncertainty rather than forcing the Student to imitate an arbitrary argmax.

Do not regress raw numerical logits with uncalibrated MSE by default. Logit scale is partly a teacher-calibration property; probability/distribution distillation is the cleaner semantic baseline.

## Auxiliary-head ablation matrix

Keep the experiment interpretable. At minimum compare:

| Variant | Boundary-now | Readiness horizons | Section distribution |
|---|---:|---:|---:|
| Baseline | — | — | — |
| B | ✓ | — | — |
| R | — | ✓ | — |
| F | — | — | ✓ |
| R+F | — | ✓ | ✓ |
| B+R+F | ✓ | ✓ | ✓ |

If data/compute is limited, prioritize:

```text
Baseline -> R -> F -> R+F
```

because transition readiness is the primary hypothesis and section-function distillation is secondary.

Do not add more SongFormer-derived heads merely because the logits make them easy to define.

## What success means

Auxiliary prediction accuracy is **not** the primary success criterion.

For each retained head ask:

> Does forcing the causal shared trunk to represent this teacher knowledge make the Player act more musically correctly?

Measure downstream behavior such as:

- earlier useful preparation before structural changes without premature commitment;
- improved native Shore-crossing timing / deadline behavior from #81/#87;
- fewer late/frantic corrections immediately after a boundary;
- better-timed Julia zoom, rotation, and palette evolution through ordinary ViewControls;
- no degradation of non-transition motion/presentation;
- improved eyes-on musical coherence on held-out songs;
- generalization beyond the songs used to fit/tune the auxiliary losses.

A head that predicts SongFormer well but does not improve Player behavior should be dropped.

## Keep auxiliary supervision distinct from behavioral loss

This ticket owns the question:

```text
Can SongFormer knowledge improve h_t through representation-shaping supervision?
```

It should **not** simultaneously answer the separate question:

```text
Should SongFormer annotations directly score/punish the Player's resulting behavior?
```

Direct transition-timing / crossing behavior objectives belong with #81/#87 or a later explicit reward ticket.

Keeping these paths separate lets us determine whether an improvement comes from a better internal causal representation or simply from directly rewarding behavior against an offline oracle.

Likewise, SongFormer structure may later contribute to offline weak ColorIntent labels under #95, but color-rule design is not part of this ticket.

## Relationship to #103

#103 evaluates broader self-hosted machine-listening teachers for perceptual targets/embeddings such as arousal, texture, tension, brightness, etc.

#102 is narrower:

```text
SongFormer -> structural boundary/function teacher -> causal auxiliary distillation.
```

The two may eventually share a simple offline-teacher artifact convention, but do not merge their research questions or require a general framework before either has produced useful evidence.

## Relationship to #104/#108/#109

- #104 owns the common causal temporal trunk and the generic auxiliary-head attachment/export contract.
- #108 owns the versioned runtime PlayerObservation.
- #109 owns causal runtime musical perception.
- This ticket may add **training targets and training-only heads**, but it must not add future-aware SongFormer values to PlayerObservation or AudioActivation.
- Removing an auxiliary head after training must not alter the product Controls-v2 contract.

## Causality / leakage tests

Mechanically verify:

- Student inputs at time `t` contain no SongFormer output and no future audio;
- future-aware teacher values are used only on the target/loss side;
- changing future audio may change the teacher target but cannot change the already-constructed causal Student input at `t`;
- normalization/calibration parameters for teacher targets are fit on training folds only where applicable;
- overlapping windows from one song/performance do not leak between train and held-out evaluation;
- teacher timeline alignment cannot be identified through padding, end-of-song, or resampling artifacts;
- auxiliary outputs are absent from or explicitly ignored by the production export path.

## Human / held-out evaluation

Do not evaluate success solely by agreement with SongFormer. That would only prove distillation.

Retain held-out eyes-on evaluation where the Domain Expert judges whether the Player:

```text
prepares appropriately
commits at sensible times
holds when a section is stable
changes presentation with restraint
avoids frantic/post-hoc correction
```

SongFormer is scalable supervision, not the final definition of musical truth.

## Acceptance / resolution

Close this research ticket with:

- [ ] dense SongFormer boundary/function teacher outputs are preserved with version/time/provenance metadata;
- [ ] teacher timelines align deterministically to canonical Player training ticks without entering runtime inputs;
- [ ] boundary-now distillation receives an explicit keep/drop verdict;
- [ ] one small transition-readiness horizon family is defined and tested causally;
- [ ] soft section/function distillation receives an explicit keep/drop verdict;
- [ ] teacher uncertainty/confidence is preserved rather than collapsed blindly to hard labels;
- [ ] auxiliary-head ablations compare against a no-SongFormer baseline under matched conditions;
- [ ] every retained head improves actual downstream Player behavior enough to justify itself, not merely its own prediction metric;
- [ ] held-out eyes-on evaluation is reported separately from teacher agreement;
- [ ] SongFormer and its future-aware outputs are absent from the live PlayerObservation/browser runtime;
- [ ] named section classes do not become required runtime state;
- [ ] auxiliary outputs are discarded at inference or retained only as explicitly optional diagnostics;
- [ ] direct teacher-based behavior losses and ColorIntent heuristics remain outside this issue.

## Non-goals

- Do not run SongFormer in the browser/live runtime.
- Do not feed SongFormer logits/probabilities/labels into runtime PlayerObservation.
- Do not hard-code `transition_readiness` in `runtime-core`.
- Do not require the Player to know symbolic verse/chorus/bridge labels at inference.
- Do not use raw teacher-logit MSE as the unquestioned distillation objective.
- Do not add a large symbolic music-theory ontology.
- Do not bypass PlayerPolicy to directly trigger Shore crossings, Julia zoom, rotation, palette changes, or Physics forces.
- Do not combine this representation-shaping experiment with direct transition behavior reward in a way that makes causal attribution impossible.
- Do not treat SongFormer as ground truth when human judgment disagrees.
